### PR TITLE
Add DDEV integration

### PR DESCRIPTION
**Pull request by martin-helmich**
This PR adds a simple integration for [DDEV](https://ddev.readthedocs.io).

This integration consists of a few simple building blocks:

- The `mw ddev render-config` command may be used to create a project-/app-specific configuration file
- The `mw ddev init` command initializes a DDEV project in the local directory, whose configuration matches the 
specified mittwald app
- External to this repository, there is also the [mittwald/ddev addon](https://github.com/mittwald/ddev), which adds support for pushing+pulling code from+to a remote project

### CODE REVIEW

**Comment by martin-helmich** [comment]
Blocked by https://github.com/mittwald/cli/pull/225, to enable automated testing.

File: src/commands/ddev/init.tsx [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    ]);
+  }
+
+  private async determineProjectName(r: ProcessRenderer): Promise<string> {
+    const { "project-name": projectName } = this.flags;
+    if (projectName) {
+      return projectName;
+    }
+
+    return await r.addInput("Enter the project name", false);
```
Comment (LukasFritzeDev) [inline src/commands/ddev/init.tsx:109]: Improvement idea (maybe another PR): Use Dirname as default value. So if an empty name is given, the directory name is used.

File: src/lib/context_ddev.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  }
+
+  /**
+   * Find the .ddev directory in the current working directory or any of its
+   * parent directories.
+   */
+  private async findDDEVConfigDir(): Promise<string | undefined> {
+    let currentDir = cwd();
+    while (currentDir !== "/") {
+      const ddevDir = path.join(currentDir, ".ddev", "config.yaml");
```
Comment (LukasFritzeDev) [inline src/lib/context_ddev.ts:79]: here it is
```ts
const ddevDir = path.join(currentDir, ".ddev", "config.yaml");
```

in `commands/ddev/init.tsx` it is:
```ts
const configFile = path.join(".ddev", "config.mittwald.yaml");
```

Can this be a problem, that in one case the `currentDir` is considered and in the other case it is not?

File: src/lib/ddev/config_builder.ts [hunk 3]
```diff
@@ -0,0 +1,209 @@
...
+        inst.appVersion.desired,
+      );
+
+      const [major] = version.externalVersion.split(".");
+      return `drupal${major}`;
+    }
+
+    throw new Error(
+      "Automatic project type detection failed. Please specify the project type manually by setting the `--override-type` flag.",
+    );
```
Comment (LukasFritzeDev) [inline src/lib/ddev/config_builder.ts:95]: Maybe use a `switch .. case` here

File: src/lib/ddev/config_builder.ts [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+    systemSoftwareVersions: SystemSoftwareVersions,
+  ): string | undefined {
+    if (!("php" in systemSoftwareVersions)) {
+      return undefined;
+    }
+
+    const version = systemSoftwareVersions["php"];
+    const [major, minor] = version.split(".");
+
+    return `${major}.${minor}`;
```
Comment (LukasFritzeDev) [inline src/lib/ddev/config_builder.ts:129]: Split on `.` and the join with `.`? 🤔 or is this to drop the patch level? 

Could be more intuitive if extracted `removePatchLevel(version)` 🤷

File: src/rendering/process/process_exec.ts [hunk 5]
```diff
@@ -0,0 +1,39 @@
+import { spawn } from "child_process";
+import { ProcessRenderer } from "./process.js";
+import { ReactNode } from "react";
+
+export async function spawnInProcess(
+  p: ProcessRenderer,
```
Comment (LukasFritzeDev) [inline src/rendering/process/process_exec.ts:6]: ahhh … above (in the command files the Renderer is always named `r`, here it’s suddenly `p`

File: src/rendering/process/process_exec.ts [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+    }
+  };
+
+  child.stdout.on("data", appendOutput);
+  child.stderr.on("data", appendOutput);
+  child.on("exit", (code) => {
+    if (code === 0) {
+      step.complete();
+    } else {
+      step.error(new Error(`rsync exited with code ${code}`));
```
Comment (LukasFritzeDev) [inline src/rendering/process/process_exec.ts:34]: Do you think only `rsync` might fail? 🤔 😜 

```suggestion
      step.error(new Error(`${cmd} exited with code ${code}`));
```

**Review by LukasFritzeDev — CHANGES_REQUESTED** [review]

File: src/commands/ddev/init.tsx (hunk 1 above)
Comment (martin-helmich) [reply src/commands/ddev/init.tsx:109]: Partially unrelated, but that reminds me: I actually wanted to add a lookup to re-use the _existing_ project name, when the command is run in a directory where a DDEV project is already initialized...

File: src/rendering/process/process_exec.ts (hunk 6 above)
Comment (martin-helmich) [reply src/rendering/process/process_exec.ts:34]: Yep, that's an oversight from refactoring. Good catch. :+1:

File: src/lib/ddev/config_builder.ts (hunk 4 above)
Comment (martin-helmich) [reply src/lib/ddev/config_builder.ts]: > or is this to drop the patch level?

Exactly this; refactoring incoming...

File: src/lib/ddev/config_builder.ts (hunk 3 above)
Comment (martin-helmich) [reply src/lib/ddev/config_builder.ts]: Incoming.

File: src/rendering/process/process_exec.ts (hunk 5 above)
Comment (martin-helmich) [reply src/rendering/process/process_exec.ts]: Fixed.

File: src/lib/context_ddev.ts (hunk 2 above)
Comment (martin-helmich) [reply src/lib/context_ddev.ts:79]: These are different use cases: The `ddev init` command is supposed to initialize a DDEV project in the current working directory of the user's shell; whereas the `DDEVContextProvider` is supposed to traverse the directory tree upwards to look for a ddev config dir _anywhere_ up the filesystem tree.

File: src/commands/ddev/init.tsx (hunk 1 above)
Comment (martin-helmich) [reply src/commands/ddev/init.tsx:116]: Regarding the default value: I wanted to mirror DDEV's behaviour here (which also prompts the user if no name is specified), as to not break anyone's expectations.

Comment (LukasFritzeDev) [reply src/commands/ddev/init.tsx:118]: yes, I do agree with the behaviour to ask for a name if none is given. But what happens if I just hit enter when the input query is shown? Does it set an empty name? In this case the dirname could be a sensible fallback, if also showed as default.

**Review by LukasFritzeDev — APPROVED** [review]

File: src/commands/ddev/init.tsx (hunk 1 above)
Comment (martin-helmich) [reply src/commands/ddev/init.tsx:118]: Created #230 as follow-up

**Pull request merged**